File: modules/mediapipe-llm/android/src/main/java/expo/modules/mediapipellm/MediaPipeLLMModule.kt
package expo.modules.mediapipellm

import java.io.File
import org.json.JSONArray
import com.google.mediapipe.tasks.genai.llminference.LlmInference
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MediaPipeLLMModule : Module() {
  private var configuredModelId: String = "gemma2b_mediapipe"
  private var modelPath: String = "/data/local/tmp/llm/gemma2b.task"
  private var llmInference: LlmInference? = null

  override fun definition() = ModuleDefinition {
    Name("MediaPipeLLM")

    Function("isAvailable") {
      return@Function true
    }

    Function("configureModel") { modelId: String, configuredPath: String? ->
      configuredModelId = modelId
      if (!configuredPath.isNullOrBlank()) {
        modelPath = configuredPath
      }
      llmInference?.close()
      llmInference = null
    }

    Function("setModelPath") { configuredPath: String ->
      modelPath = configuredPath
      llmInference?.close()
      llmInference = null
    }

    Function("getModelPath") {
      return@Function modelPath
    }

    Function("isModelReady") {
      return@Function File(modelPath).exists()
    }

    Function("getConfiguredModelId") {
      return@Function configuredModelId
    }

    AsyncFunction("generate") { messagesJson: String, temperature: Double?, maxTokens: Int? ->
      if (!File(modelPath).exists()) {
        throw IllegalStateException("Model file not found at path: $modelPath")
      }

      val inference = getOrCreateInference(temperature, maxTokens)
      val userPrompt = extractLastUserMessage(messagesJson)
      return@AsyncFunction inference.generateResponse(userPrompt)
    }
  }

  private fun getOrCreateInference(temperature: Double?, maxTokens: Int?): LlmInference {
    llmInference?.let { return it }
    val context = requireNotNull(appContext.reactContext)

    val builder = LlmInference.LlmInferenceOptions.builder()
      .setModelPath(modelPath)
      .setMaxTokens(maxTokens ?: 512)
      .setTemperature((temperature ?: 0.2).toFloat())
      .setTopK(40)
      .setRandomSeed(101)

    val options = builder.build()
    val created = LlmInference.createFromOptions(context, options)
    llmInference = created
    return created
  }

  private fun extractLastUserMessage(messagesJson: String): String {
    return try {
      val arr = JSONArray(messagesJson)
      for (i in arr.length() - 1 downTo 0) {
        val item = arr.optJSONObject(i) ?: continue
        if (item.optString("role") == "user") {
          return item.optString("content", "")
        }
      }
      ""
    } catch (_: Throwable) {
      ""
    }
  }
}

File: modules/mediapipe-llm/ios/MediaPipeLLMModule.swift
import ExpoModulesCore
#if canImport(MediaPipeTasksGenai)
import MediaPipeTasksGenai
#endif

public final class MediaPipeLLMModule: Module {
  private var configuredModelId: String = "gemma2b_mediapipe"
  private var modelPath: String = ""
#if canImport(MediaPipeTasksGenai)
  private var llmInference: LlmInference?
#endif

  public func definition() -> ModuleDefinition {
    Name("MediaPipeLLM")

    Function("isAvailable") { () -> Bool in
#if canImport(MediaPipeTasksGenai)
      true
#else
      false
#endif
    }

    Function("configureModel") { (modelId: String, configuredPath: String?) -> Void in
      self.configuredModelId = modelId
      if let configuredPath, !configuredPath.isEmpty {
        self.modelPath = configuredPath
      }
#if canImport(MediaPipeTasksGenai)
      self.llmInference = nil
#endif
    }

    Function("setModelPath") { (configuredPath: String) -> Void in
      self.modelPath = configuredPath
#if canImport(MediaPipeTasksGenai)
      self.llmInference = nil
#endif
    }

    Function("getModelPath") { () -> String in
      self.modelPath
    }

    Function("isModelReady") { () -> Bool in
      guard !self.modelPath.isEmpty else { return false }
      return FileManager.default.fileExists(atPath: self.modelPath)
    }

    Function("getConfiguredModelId") { () -> String in
      self.configuredModelId
    }

    AsyncFunction("generate") { (messagesJson: String, temperature: Double?, maxTokens: Int?) -> String in
#if canImport(MediaPipeTasksGenai)
      guard !self.modelPath.isEmpty, FileManager.default.fileExists(atPath: self.modelPath) else {
        throw NSError(domain: "MediaPipeLLM", code: 1, userInfo: [NSLocalizedDescriptionKey: "Model file not found at path: \(self.modelPath)"])
      }

      let userPrompt = Self.extractLastUserMessage(messagesJson: messagesJson)
      let inference = try self.getOrCreateInference(temperature: temperature, maxTokens: maxTokens)
      return try inference.generateResponse(inputText: userPrompt)
#else
      let userPrompt = Self.extractLastUserMessage(messagesJson: messagesJson)
      let temp = temperature ?? 0.2
      let max = maxTokens ?? 512
      return "[MediaPipe unavailable] Model=\(self.configuredModelId) temp=\(temp) maxTokens=\(max). Add MediaPipeTasksGenAI pods. Last user message: \(userPrompt)"
#endif
    }
  }

#if canImport(MediaPipeTasksGenai)
  private func getOrCreateInference(temperature: Double?, maxTokens: Int?) throws -> LlmInference {
    if let llmInference {
      return llmInference
    }

    let options = LlmInferenceOptions()
    options.baseOptions.modelPath = modelPath
    options.maxTokens = maxTokens ?? 512
    options.topk = 40
    options.temperature = Float(temperature ?? 0.2)
    options.randomSeed = 101

    let created = try LlmInference(options: options)
    self.llmInference = created
    return created
  }
#endif

  private static func extractLastUserMessage(messagesJson: String) -> String {
    guard
      let data = messagesJson.data(using: .utf8),
      let list = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
    else {
      return ""
    }

    for item in list.reversed() {
      guard let role = item["role"] as? String, role == "user" else { continue }
      return item["content"] as? String ?? ""
    }
    return ""
  }
}
